### ISSUE TITLE
Add official word-graph display support

### ISSUE THREAD

**Issue by ampli** [opencog/link-grammar#741]
The main reason why I didn't propose to include a default support for word-graph display is that  the library then includes code that issues exec/popen, a thing that potentially makes it "less secure".
The program is `dot`, and on Windows also `PhotoViewer`, and `link-parser` can invoke them by using an argument or a `!` command-line.

However, many programs, e.g. `git`, include code to execute some programs (like `diff` or an editor), and even arbitrary programs controlled by environment variables like `EDITOR` and `PAGER`, that are searched in PATH. So there is actually nothing special in allowing a similar thing for `link-parser` (or any parser script that implements reading commands from its input).
In addition, if one for some reason wants to use an unmodified `link-parser` as a web service, the common practice is to discard environment variables (that should not be controllable by the Internet user anyway) and filter possible input (to exclude lines starting with `!` in that case) so there is also no problem here.

Another reason is that `dot` used to include double-free errors that randomly caused it to crash. But this may not a good enough reason. I also checked it with version 2.40.1 (Fedora 27) and it didn't crash on any sentence in the ru/en batches, so maybe this problem got fixed in recent versions.

My proposal:
1. Enable word-graph display by default in `configure`.
2. Add a String user variable to control it instead of the current `!test=wg:flags`. The name `wordgraph-display` is too long for displaying in `!variables`, so maybe it can be `wg-display` (e.g.: `!wg-display=s`).

**Comment by linas** [comment]
I guess that's OK. Does this require new compile-time dependencies?

**Comment by ampli** [comment]
Not as far as I know.
(For runtime it needs the `graphviz` package, so I guess packagers would want to add it to the recommended packages list to be installed. But they may not know about that possibility and I don't know where is the best place to mention that.)

**Comment by ampli** [comment]
Implementations details (still unfinished):
1. Change `void wordgraph_show((Sentence sent, const char *modestr)` to
`bool sentence_display_wordgraph(Sentence sent, const char *modestr)` and define it as an API.
The intention is that it return success or failure.
2. In link parse, define user variable. I mentioned `wg-display`. But maybe it should be `wordgraph` (like `constituents` and not `conts-print`...).
3. In configure, enable wordgraph-display by default.

**Comment by ampli** [comment]
The link-parser help text (in case you would like that I will reword it) is:
Display sentence word-graph

(I still don't know how to officially call what I originally named Wordgraph.)

**Comment by linas** [comment]
Calling it wordgraph seems OK. Alternatives might be "morphgraph" but that is a mouthful. I like the article https://en.wikipedia.org/wiki/Catena_(linguistics) but it has nothing to do with this, not really.

tokengraph is another possibility. tokenstream. tokenchain. toeknseq token-sequence, tokenunit graph. tokenization graph. 

segment graph, seggraph, wordseg, token segmentation  

I think I like "tokengraph" its more euphonious that segmentgraph or seggraph, and is less generic than wordgraph.  The tokens could be words, morphemes, or something else.

https://en.wikipedia.org/wiki/Text_segmentation

**Comment by ampli** [comment]
So how we proceed in changing its name to "tokengraph"?
It is tedious to change all the existing code/comments/documents...
If we decide to make this terminology change, the most important immediate changes are:
1. The new API function sentence_display_wordgraph().
2. link-parser's new user variable `wordgraph` (including the help text).
3. The option name in `configure`.
4. The new section in `tokenize/README.md`.

This can be done before or after applying the current PRs ("before" means a need to force-push them).
Or is there another way?

**Comment by linas** [comment]
Yes, it's tedious to make the change.  Would you care to surf the net, and see if any other NLP projects have anything comparable, and what they might call it?

Re popen: if we are walking down that route, we could also do the same for the postscript output.

**Comment by ampli** [comment]
> Re popen: if we are walking down that route, we could also do the same for the postscript output.

This is indeed possible (provided that the postscript code gets fixed to support long sentences).

BTW,  I once started to rewrite link-parser in Python (half done) and intended to use a graphical module (need to recall its name) for display. It turns out that some graphical modules have a postscript widget that can be used for that. Many interesting things can be done with a link-parser program that has a GUI (think about clicking on various items).
For the current link-parser I guess that ghostscript will be an easy route.

**Comment by ampli** [comment]
> Would you care to surf the net, and see if any other NLP projects have anything comparable, and what they might call it?

I will do that.

**Comment by linas** [comment]
If you wanted to write a GUI for link-parser, I would not discourage you.  People like GUI's.

**Comment by linas** [comment]
Heh. The definition in the "token graph" paper is interesting. Its kind of hard to understand at first, .. but interesting.  

Here's what I'm tempted to say: "so what if its unrelated; it is a rare usage, they invented the term token graph on 2011 and its  not in any textbooks, and there are 5 citations total for the article. So its OK if we use the same name for something different." 

I think that's OK. And its not entirely unrelated: they're describing different "alternatives" for placing tokens on a graph. We're describing "alternatives", also.  Graph theory is weird and hard.

Or we can tall it a segment graph, or a token chain.

**Comment by ampli** [comment]
Isn't it just "token alternatives graph"?

**Comment by linas** [comment]
That will do.  However, in the code we have "wg", do you want to replace that with "tag"? We have a file called "wordgraph.c" do you want to rename this to "token_alternatives_graph.c" ?  I assume the answer is no; thus the conundrum.

**Comment by ampli** [comment]
Indeed "token alternatives graph" is somewhat cumbersome.

Maybe we can refer to it as a "word split graph" and use "wordgraph" as a short for that... This will take less changes...

(Generally it should even be "word split/combine graph" as general tokenization sometimes may need to combine words.)

We just need to find out how to call it, and then plan how to make the rename.

**Comment by linas** [comment]
or we can just leave the name unchanged.  I was looking to change the name only so that some hypothetical, non-existent outsider/newcomer could look at it, and instantly understand that "hey, this is where the morphemes are". 

Naming things is hard.

**Comment by ampli** [comment]
So to make it clearer, I will just add in the relevant README and in a source code comment something like `word split graph, AKA "wordgraph"`.

**Comment by linas** [comment]
sure.

**Comment by ampli** [comment]
The docs now include "wordgraph" in several versions:
- With and w/o a dash (word-graph vs wordgraph).
- Capitalized or not.

Options:
- Leave it as it.
- Change it to a consistent name. Which one?

**Comment by ampli** [comment]
I guess the term should be "word-graph". It is already used as such in many places, including the help texts of `link-parser` (but the user-variable is called `wordgraph`). However, in the manual and some other places it is still "wordgraph". So I guess everything should be changed to "wordgraph", and maybe the user variable should be added an internal synonym `word-graph` (an addition of one table line).
